python/cog/server/worker.py

```diff
@@ -221,13 +223,18 @@ def _predict(self, payload: Dict[str, Any]) -> None:
             result = predict(**payload)
 
             if result:
-                if isinstance(result, types.GeneratorType):
-                    self._events.send(PredictionOutputType(multi=True))
-                    for r in result:
-                        self._events.send(PredictionOutput(payload=make_encodeable(r)))
-                else:
-                    self._events.send(PredictionOutputType(multi=False))
-                    self._events.send(PredictionOutput(payload=make_encodeable(result)))
+                with self._events_lock:
+                    if isinstance(result, types.GeneratorType):
+                        self._events.send(PredictionOutputType(multi=True))
+                        for r in result:
```

Comment:
This is holding the lock for wayyyyyy too long. If the result is generator we absolutely should not be holding the lock for the entire `for r in result` loop.

I wonder if this is related to the occasional customer report of not getting logs?